src/laser/dipole_beam.rs

```diff
@@ -0,0 +1,99 @@
+extern crate specs;
+use crate::constant;
+use serde::{Deserialize, Serialize};
+use specs::{
+    Component, Entities, HashMapStorage, Join, LazyUpdate, Read, ReadStorage, System, WriteStorage,
+};
+
+/// A component marking the entity as laser beam for dipole forces and
+/// holding properties of the light
+#[derive(Deserialize, Serialize, Clone, Copy)]
+pub struct DipoleLight {
+    ///wavelength of the laser light in SI units of m.
+    pub wavelength: f64,
+}
+
+impl DipoleLight {
+    /// Frequency of the dipole light in units of Hz
+    pub fn frequency(&self) -> f64 {
+        constant::C / self.wavelength
+    }
+
+    /// Wavenumber of the dipole light, in units of 2pi/m
+    pub fn wavenumber(&self) -> f64 {
+        2.0 * constant::PI / self.wavelength
+    }
+}
+
+impl Component for DipoleLight {
+    type Storage = HashMapStorage<Self>;
+}
+
+/// An index that uniquely identifies this dipole light in the interaction list for each atom.
+/// The index value corresponds to the position of each dipole light in the per-atom interaction list array.
+///
+/// Default `DipoleLightIndex`s are created with `initiated: false`.
+/// Once the index is set, initiated is set to true.
+/// This is used to detect if all lasers in the simulation are correctly indexed, in case new lasers are added.
+#[derive(Clone, Copy)]
+pub struct DipoleLightIndex {
```

Comment:
I won't make this change now, but recording this thought for the future.

To some extent the cooling light and dipole light mirror each other. They both have a components (eg `CoolingLight`/`DipoleLight`) and both have an `Index` component. It is possible that we could improve the code by defining a generic `Light<>` and `LightIndex<>` component, which supports `Light<Cooling>` etc.

Alternatively, we could add a define macro that generates the boilerplate code for each of these, eg generating `...Light` and `...LightIndex`, and the relevant systems. Then you would invoke the macro eg `light(Cooling)` and `light(Dipole)` and the rest would be made automatically. We should avoid copy and paste where possible.

Reply:
(Leaving the comment for you to resolve when you've read it)

Reply:
good thought!